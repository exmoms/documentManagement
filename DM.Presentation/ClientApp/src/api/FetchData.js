/**
 * @file this file includes functions that fetch data from API
 */

// Generic Delete Method
export async function deleteFromApi(url) {
  let res = await fetch(url, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  });
  return res;
}

// Get MetaData Models
export async function fetchData(url) {
  let object = {};
  /*
   * send the request through a CORS proxy
   * https://stackoverflow.com/questions/43262121/trying-to-use-fetch-and-pass-in-mode-no-cors/43268098#43268098
   */
  await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  })
    .then((response) => response.json())
    .then(
      (data) => (object = data),
      (error) => console.log(error)
    );
  return object;
}
