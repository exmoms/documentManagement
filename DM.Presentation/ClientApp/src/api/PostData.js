/**
 * @file This file includes functions that help to post data to a specific API
 * @version 1.0
 */

/**
 * Post new document data.
 * Upload form-data, scanned copy of doc and attachments.
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @param {*} documentData
 * @param {*} images
 * @param {*} attachments
 * @param {*} url
 * @param {*} option its values either "Add" for add new document version or "Update" to update exsiting document Version
 */
export async function postDocumentData(
  documentData,
  scannedPages,
  attachments,
  url = "",
  option
) {
  var formData = new FormData();

  if (process.env.NODE_ENV === "development") {
    console.log(documentData);
    console.log(attachments);
    console.log(scannedPages);
  }

  switch (option) {
    case "Add":
      formData.append("Document", JSON.stringify(documentData));
      break;
    case "Update":
      formData.append("Version", JSON.stringify(documentData));
      break;
    default:
  }

  attachments.forEach((attachment) => {
    formData.append("Attachments", attachment);
  });
  scannedPages.forEach((scannedPage) => {
    formData.append("Scans", scannedPage);
  });

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    headers: {
      Accept: "application/json",
      // "Content-Type": "application/json",
    },
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Attachment Data.
 * Upload attachment json data + attachment file  .
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} payload
 * @param {*} attachmentFile
 * @param {*} url
 */
export async function postAttachmentData(payload, attachmentFile, url = "") {
  var formData = new FormData();

  formData.append("Attachment", JSON.stringify(payload));
  formData.append("FileAttachment", attachmentFile);

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Request To API
 * Send post request to api with  payload  and return a response.
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} url The api url
 * @param {*} payload js object which will be sent as request body.
 */
export async function postDataToAPI(url, payload = null, method = "POST") {

  const response = await fetch(url, {
    method: method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  return response;
}
