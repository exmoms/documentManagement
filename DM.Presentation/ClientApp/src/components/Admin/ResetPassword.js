import Avatar from "@material-ui/core/Avatar";
import Button from "@material-ui/core/Button";
import Container from "@material-ui/core/Container";
import CssBaseline from "@material-ui/core/CssBaseline";
import FormControl from "@material-ui/core/FormControl";
import Grid from "@material-ui/core/Grid";
import IconButton from "@material-ui/core/IconButton";
import InputAdornment from "@material-ui/core/InputAdornment";
import InputLabel from "@material-ui/core/InputLabel";
import LinearProgress from "@material-ui/core/LinearProgress";
import OutlinedInput from "@material-ui/core/OutlinedInput";
import { makeStyles } from "@material-ui/core/styles";
import TextField from "@material-ui/core/TextField";
import Typography from "@material-ui/core/Typography";
import LockOutlinedIcon from "@material-ui/icons/LockOutlined";
import Visibility from "@material-ui/icons/Visibility";
import VisibilityOff from "@material-ui/icons/VisibilityOff";
import clsx from "clsx";
import React, { useState } from "react";
import { useTranslation } from "react-i18next";
import { Redirect } from "react-router-dom";
import { postDataToAPI } from "../../api/PostData";
import { Notification } from "../Admin/Notifications";

const useStyles = makeStyles((theme) => ({
  paper: {
    marginTop: theme.spacing(8),
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
  },
  form: {
    width: "100%", // Fix IE 11 issue.
    marginTop: theme.spacing(3),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
  margin: {
    margin: theme.spacing(1),
  },
  textField: {
    width: "100%",
  },
}));

export default function ResetsPassword() {
  const { t } = useTranslation();
  const classes = useStyles();

  const [loading, setLoading] = useState(false);
  const [sent_successfuly, setSent_successfuly] = useState(false);
  const [confirmPassword, setConfirmPassword] = useState("");
  const [values, setValues] = useState({
    password: "",
    showPassword: false,
  });

  const [open, setOpen] = useState(false);
  const [errorMessage, seterrorMessage] = useState([]);
  const [error, seterror] = useState(false);

  const onchange = (event) => {
    if (event.target.name === "confirmPassword") {
      setConfirmPassword(event.target.value);
    }
  };

  const handelSubmit = async (event) => {
    event.preventDefault();
    setLoading(true);
    let token = window.location.href.substring(
      window.location.href.indexOf("token=") + 6,
      window.location.href.indexOf("&email=")
    );
    let email = window.location.href.substring(
      window.location.href.indexOf("&email=") + 7
    );

    const post_file = {
      token: token,
      email: email,
      password: values.password,
      confirmPassword: confirmPassword,
    };

    if (values.password !== confirmPassword) {
      setOpen(true);
      seterror(true);
      seterrorMessage([t("password mismatch")]);
    } else {
      let response = await postDataToAPI("/api/user/ResetPassword", post_file);
      if (response.ok) {
        setOpen(true);
        setSent_successfuly(true);
        setLoading(false);
      } else {
        let json = await response.json();
        setOpen(true);
        seterror(true);
        seterrorMessage(json.error);
        setLoading(false);
      }
    }
  };

  const handleChange = (prop) => (event) => {
    setValues({ ...values, [prop]: event.target.value });
  };

  const handleClickShowPassword = () => {
    setValues({ ...values, showPassword: !values.showPassword });
  };

  const handleMouseDownPassword = (event) => {
    event.preventDefault();
  };

  if (sent_successfuly === true) {
    return <Redirect to="/" />;
  } else {
    return (
      <Container component="main" maxWidth="xs">
        <CssBaseline />
        <div className={classes.paper}>
          <Avatar className={classes.avatar}>
            <LockOutlinedIcon />
          </Avatar>
          <Typography component="h1" variant="h5">
            {t("resetpassword")}
          </Typography>
          <form className={classes.form} onSubmit={handelSubmit} id="formSubmit">
            <Grid container spacing={2}>
              <FormControl
                className={clsx(classes.margin, classes.textField)}
                variant="outlined"
              >
                <InputLabel htmlFor="outlined-adornment-password">
                  {t("password")}
                </InputLabel>
                <OutlinedInput
                  required
                  id="outlined-adornment-password"
                  type={values.showPassword ? "text" : "password"}
                  value={values.password}
                  onChange={handleChange("password")}
                  endAdornment={
                    <InputAdornment position="end">
                      <IconButton
                        aria-label="toggle password visibility"
                        onClick={handleClickShowPassword}
                        onMouseDown={handleMouseDownPassword}
                        edge="end"
                      >
                        {values.showPassword ? (
                          <Visibility />
                        ) : (
                          <VisibilityOff />
                        )}
                      </IconButton>
                    </InputAdornment>
                  }
                  labelWidth={70}
                />
              </FormControl>

              <Grid item xs={12}>
                <TextField
                  variant="outlined"
                  required
                  fullWidth
                  onChange={onchange}
                  name="confirmPassword"
                  label="confirm Password"
                  type="password"
                  id="confirmPassword"
                  autoComplete="current-password"
                />
              </Grid>
            </Grid>
            {loading && <LinearProgress color="secondary" />}
            <Button
              type="submit"
              fullWidth
              variant="contained"
              color="primary"
              className={classes.submit}
            >
              {t("reset")}
            </Button>
          </form>
          <Notification
          id="notification"
            open={open} // must be a "state" to show notification message [true or false]
            setOpen={setOpen} // this function changes the state "open" value.
            error={error} // if error message put [true] else for success put [false]
            errorMessage={errorMessage} // the recived error message (array of string) from server.
          />
        </div>
      </Container>
    );
  }
}
