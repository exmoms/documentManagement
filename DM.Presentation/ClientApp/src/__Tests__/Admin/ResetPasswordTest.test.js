
import React from "react";
import ReactDom from "react-dom";
import ResetPassword from "../../components/Admin/ResetPassword";
import { cleanup } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { shallow , mount} from "enzyme";
import toJSON from "enzyme-to-json";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { Link } from "react-router-dom";
import { Suspense } from "react";
import * as FetchMock from "../../api/FetchData";
import * as PostDataMock from "../../api/PostData";
import * as ContextAuth from "../../context/auth";
import MenuItem from "@material-ui/core/MenuItem";
import { act } from "react-dom/test-utils";
import List from "@material-ui/core/List";
import {Redirect } from "react-router-dom";
import Typography from "@material-ui/core/Typography";
import LinearProgress from "@material-ui/core/LinearProgress";
import Visibility from "@material-ui/icons/Visibility";
import VisibilityOff from "@material-ui/icons/VisibilityOff";
import IconButton from "@material-ui/core/IconButton";
afterEach(cleanup);
describe("ResetPassword", () => {
   afterEach(() => {
    jest.clearAllMocks();
  });
  test("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDom.render(
        <ResetPassword/>,div
    );
  });

  test("matches snapshot", () => {
    const wrapper = shallow(
        <ResetPassword/>
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  
  test("intialization when sent_successfuly is false", () => {
    let  wrapper = mount(<ResetPassword/>);
    expect(wrapper.find("#outlined-adornment-password").at(0).props().value).toBe("");
    expect(wrapper.find("#outlined-adornment-password").at(0).props().type).toBe("password");
    expect(wrapper.find(Visibility).length).toBe(0);
    expect(wrapper.find(VisibilityOff).length).toBe(1);
    expect(wrapper.find(LinearProgress).length).toBe(0);
  });

  test("handleClickShowPassword", () => {
    let  wrapper = mount(<ResetPassword />);
    act(() => {
        wrapper
          .find(IconButton)
          .at(0)
          .props()
          .onClick();
      });
      wrapper.render();
        wrapper.update();
    expect(wrapper.find("#outlined-adornment-password").at(0).props().value).toBe("");
    expect(wrapper.find("#outlined-adornment-password").at(0).props().type).toBe("text");
    expect(wrapper.find(Visibility).length).toBe(1);
    expect(wrapper.find(VisibilityOff).length).toBe(0);
    expect(wrapper.find(LinearProgress).length).toBe(0);
  });

  test("handleMouseDownPassword", () => {
    let  wrapper = mount(<ResetPassword />);
    const preventDefaultMock = jest.fn();
    act(() => {
        wrapper
          .find(IconButton)
          .at(0)
          .props()
          .onMouseDown({ preventDefault:preventDefaultMock});
      });
      wrapper.render();
        wrapper.update();
        expect(preventDefaultMock).toHaveBeenCalledTimes(1);
        expect(wrapper.find("#outlined-adornment-password").at(0).props().value).toBe("");
    expect(wrapper.find("#outlined-adornment-password").at(0).props().type).toBe("password");
    expect(wrapper.find(Visibility).length).toBe(0);
    expect(wrapper.find(VisibilityOff).length).toBe(1);
    expect(wrapper.find(LinearProgress).length).toBe(0);
  });

  test("handleChange", () => {
    let  wrapper = mount(<ResetPassword />);
    expect(wrapper.find("#outlined-adornment-password").at(0).props().value).toBe("");
    expect(wrapper.find("#outlined-adornment-password").at(0).props().type).toBe("password");
    expect(wrapper.find(Visibility).length).toBe(0);
    expect(wrapper.find(VisibilityOff).length).toBe(1);
    act(() => {
        wrapper
          .find("#outlined-adornment-password")
          .at(0)
          .props()
          .onChange({ target: { value:  "password" } });
      });
      wrapper.render();
        wrapper.update();
        expect(wrapper.find("#outlined-adornment-password").at(0).props().value).toBe("password");
        expect(wrapper.find("#outlined-adornment-password").at(0).props().type).toBe("password");
        expect(wrapper.find(Visibility).length).toBe(0);
        expect(wrapper.find(VisibilityOff).length).toBe(1);
        expect(wrapper.find(LinearProgress).length).toBe(0);
  });

  test("handelSubmit when  (values.password !== confirmPassword) ", () => {
    const preventDefaultMock = jest.fn();
    const logMock=  jest.spyOn(console, "log");
    const alertMock=  jest.spyOn(window, "alert");
    let  wrapper = mount(<ResetPassword />);
    //set the password
    act(() => {
        wrapper
          .find("#outlined-adornment-password")
          .at(0)
          .props()
          .onChange({ target: { value:  "password" } });
      });
      wrapper.render();
      wrapper.update();
    expect(wrapper.find("#outlined-adornment-password").at(0).props().value).toBe("password");
    // set confirm Password
    act(() => {
        wrapper
          .find("#confirmPassword")
          .at(0)
          .props()
          .onChange({ target: { value: "confirmPassword" ,name:"confirmPassword"} });
      });
        wrapper.update();
    act(() => {
        wrapper
          .find("#formSubmit")
          .at(0)
          .props()
          .onSubmit({ preventDefault:preventDefaultMock});
      });
      wrapper.render();
        wrapper.update();
        expect(preventDefaultMock).toHaveBeenCalledTimes(1);
        expect(wrapper.find("#outlined-adornment-password").at(0).props().value).toBe("password");
        expect(wrapper.find("#outlined-adornment-password").at(0).props().type).toBe("password");
        expect(wrapper.find(Visibility).length).toBe(0);
        expect(wrapper.find(VisibilityOff).length).toBe(1);
        expect(wrapper.find(LinearProgress).length).toBe(1);
       expect(wrapper.find("#notification").props().open).toBeTruthy(); 
      expect(wrapper.find("#notification").props().error).toBeTruthy(); 
      expect(wrapper.find("#notification").props().errorMessage).toStrictEqual(["password mismatch"]); 
  });

  test("handelSubmit when  res is ok and onchange", () => {
    const mockSuccessResponsePost ={ok :true};
    const mockJsonPromisePost = Promise.resolve(mockSuccessResponsePost);
    const PostMock = jest.spyOn(PostDataMock, "postDataToAPI").mockImplementation(() => mockJsonPromisePost);
    const preventDefaultMock = jest.fn();
    const logMock=  jest.spyOn(console, "log");
    const alertMock=  jest.spyOn(window, "alert");
    let  wrapper = mount(<Router><ResetPassword /></Router>);
    //set the password
    act(() => {
        wrapper
          .find("#outlined-adornment-password")
          .at(0)
          .props()
          .onChange({ target: { value:  "passwordA" } });
      });
      wrapper.render();
      wrapper.update();
    // set confirm Password
    act(() => {
        wrapper
          .find("#confirmPassword")
          .at(0)
          .props()
          .onChange({ target: { value: "passwordA" ,name:"confirmPassword"} });
      });
        wrapper.update();
    return mockJsonPromisePost.then(() => {
        act(() => {
            wrapper
              .find("#formSubmit")
              .at(0)
              .props()
              .onSubmit({ preventDefault:preventDefaultMock});
          });
    }).then(() => {
        wrapper.render();
        wrapper.update();
        
        expect(preventDefaultMock).toHaveBeenCalledTimes(1);
        expect(PostMock).toHaveBeenCalledTimes(1);
        expect(PostMock.mock.calls[0][0]).toBe("/api/user/ResetPassword");
        expect(PostMock.mock.calls[0][1]).toStrictEqual({"confirmPassword": "passwordA", "email": "/localhost/", "password": "passwordA", "token": "http:"});
        expect(wrapper.find(Redirect).length).toBe(1);
    });

        
  });

  test("handelSubmit when  res isn' ok and onchange", (done) => {
    let mockSuccessErrorPromise = { error: ["errorMessage"] };
    let mockSuccessError = Promise.resolve(mockSuccessErrorPromise);
    const mockSuccessResponsePost = Promise.resolve({
      ok: false,
      json: () => mockSuccessError,
    });
    const PostMock = jest
      .spyOn(PostDataMock, "postDataToAPI")
      .mockReturnValue(mockSuccessResponsePost);
    const preventDefaultMock = jest.fn();
    const logMock = jest.spyOn(console, "log");
    const alertMock = jest.spyOn(window, "alert");
    let wrapper = shallow(<ResetPassword />);
    //set the password
    act(() => {
      wrapper
        .find("#outlined-adornment-password")
        .at(0)
        .props()
        .onChange({ target: { value: "passwordA" } });
    });
    wrapper.update();
    wrapper.render();

    // set confirm Password
    act(() => {
      wrapper
        .find("#confirmPassword")
        .at(0)
        .props()
        .onChange({ target: { value: "passwordA", name: "confirmPassword" } });
    });
    wrapper.update();
    wrapper.render();

    act(() => {
      wrapper
        .find("#formSubmit")
        .at(0)
        .props()
        .onSubmit({ preventDefault: preventDefaultMock });
    });
    wrapper.update();
    wrapper.render();

    expect(preventDefaultMock).toHaveBeenCalledTimes(1);
    expect(PostMock).toHaveBeenCalledTimes(1);
    expect(PostMock.mock.calls[0][0]).toBe("/api/user/ResetPassword");
    expect(PostMock.mock.calls[0][1]).toStrictEqual({
      confirmPassword: "passwordA",
      email: "/localhost/",
      password: "passwordA",
      token: "http:",
    });
    process.nextTick(() => {

      expect(wrapper.find("#notification").props().open).toBeTruthy();
      expect(wrapper.find("#notification").props().error).toBeTruthy();
      expect(wrapper.find("#notification").props().errorMessage).toStrictEqual([
        "errorMessage",
      ]);
      done(); 
    });
  });
});
