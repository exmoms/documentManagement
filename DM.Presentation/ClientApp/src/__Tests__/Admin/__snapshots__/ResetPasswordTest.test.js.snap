// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword matches snapshot 1`] = `
<WithStyles(ForwardRef(Container))
  component="main"
  maxWidth="xs"
>
  <WithStyles(CssBaseline) />
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(Avatar))>
      <LockOutlinedIcon />
    </WithStyles(ForwardRef(Avatar))>
    <WithStyles(ForwardRef(Typography))
      component="h1"
      variant="h5"
    >
      resetpassword
    </WithStyles(ForwardRef(Typography))>
    <form
      className="makeStyles-form-2"
      id="formSubmit"
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        spacing={2}
      >
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-margin-4 makeStyles-textField-5"
          variant="outlined"
        >
          <WithStyles(ForwardRef(InputLabel))
            htmlFor="outlined-adornment-password"
          >
            password
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(OutlinedInput))
            endAdornment={
              <ForwardRef(WithStyles)
                position="end"
              >
                <ForwardRef(WithStyles)
                  aria-label="toggle password visibility"
                  edge="end"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                >
                  <UNDEFINED />
                </ForwardRef(WithStyles)>
              </ForwardRef(WithStyles)>
            }
            id="outlined-adornment-password"
            labelWidth={70}
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          />
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={12}
        >
          <WithStyles(ForwardRef(TextField))
            autoComplete="current-password"
            fullWidth={true}
            id="confirmPassword"
            label="confirm Password"
            name="confirmPassword"
            onChange={[Function]}
            required={true}
            type="password"
            variant="outlined"
          />
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Button))
        className="makeStyles-submit-3"
        color="primary"
        fullWidth={true}
        type="submit"
        variant="contained"
      >
        reset
      </WithStyles(ForwardRef(Button))>
    </form>
    <Notification
      error={false}
      errorMessage={Array []}
      id="notification"
      open={false}
      setOpen={[Function]}
    />
  </div>
</WithStyles(ForwardRef(Container))>
`;
